src/eos-updater-apply.c

```diff
@@ -73,41 +79,439 @@ get_test_osname (void)
-                EosUpdaterData *data,
-                GCancellable *cancel,
-                gboolean *out_bootversion_changed,
-                GError **error)
+cleanup_undeployed_flatpaks (GPtrArray *pending_flatpak_ref_actions,
+                             gsize      right_bound,
+                             GError   **error)
+{
+  g_autoptr(FlatpakInstallation) installation = eos_updater_get_flatpak_installation (NULL,
+                                                                                      error);
+
+  if (!installation)
+    return FALSE;
+
+  cleanup_undeployed_flatpaks_in_installation (installation,
+                                               pending_flatpak_ref_actions,
+                                               right_bound);
+
+  return TRUE;
+}
+
+static void
+async_result_cb (GObject      *obj,
+                 GAsyncResult *result,
+                 gpointer      user_data)
+{
+  GAsyncResult **result_out = user_data;
+  *result_out = g_object_ref (result);
+}
+
+/* As documented in OstreeRepoFinderResult, lower numbers indicate higher
+ * priority */
+static OstreeRepoFinderResult *
+highest_priority_repo_finder_result (OstreeRepoFinderResultv results)
+{
+  gint best_priority_score = G_MAXINT;
+  OstreeRepoFinderResult *best = NULL;
+  gsize i;
+
+  for (i = 0; results[i] != NULL; ++i)
+    {
+      if (results[i]->priority < best_priority_score)
+        {
+          best_priority_score = results[i]->priority;
+          best = results[i];
+        }
+    }
+
+  return best;
+}
+
+static gchar *
+find_first_flatpak_repo_for_collection_id (FlatpakInstallation        *installation,
+                                           const OstreeCollectionRef  *collection_ref,
+                                           GCancellable               *cancellable,
+                                           GError                    **error)
+{
+  g_autoptr(GFile) installation_directory = flatpak_installation_get_path (installation);
+  g_autoptr(GFile) repo_directory = g_file_get_child (installation_directory, "repo");
+  g_autoptr(GMainContext) context = g_main_context_new ();
+  g_autoptr(OstreeRepo) repo = ostree_repo_new (repo_directory);
+  g_autoptr(GAsyncResult) find_result = NULL;
+  g_auto(OstreeRepoFinderResultv) allocated_results = NULL;
+  OstreeCollectionRef const * collection_refs_to_fetch[2] = { collection_ref, NULL };
+
+  g_main_context_push_thread_default (context);
+
+  if (!ostree_repo_open (repo, cancellable, error))
+    return NULL;
+
+  ostree_repo_find_remotes_async (repo,
+                                  collection_refs_to_fetch,
+                                  NULL,
+                                  NULL,
+                                  NULL,
+                                  NULL,
+                                  async_result_cb,
+                                  &find_result);
+
+  while (find_result == NULL)
+    g_main_context_iteration (context, TRUE);
```

Comment:
Is there no sync version of this function, instead of manually iterating the main context here?

Reply:
Unfortunately not - it looks like this is the pattern in use elsewhere in this code as well.

Reply:
This is the correct pattern. libostree is basically always meant to be used from a worker thread.